python/taichi/testing.py

```diff
@@ -0,0 +1,97 @@
+# The load-on-request closure (LoR-C) infrastructure, credit by @archibate
+def _():
+    ############## HUMAN TOUCHABLE BEGIN ###################
+    import taichi as ti
+
+    print('[Taichi] loading test module')
+
+    import functools
+    import pytest
+
+    ## Helper functions
+    def approx(expected, **kwargs):
+        '''Tweaked pytest.approx for OpenGL low percisions'''
+        class boolean_integer:
+            def __init__(self, value):
+                self.value = value
+
+            def __eq__(self, other):
+                return bool(self.value) == bool(other)
+
+            def __ne__(self, other):
+                return bool(self.value) != bool(other)
+
+        if isinstance(expected, bool):
+            return boolean_integer(expected)
+
+        if ti.cfg.arch == ti.opengl:
+            kwargs['rel'] = max(kwargs.get('rel', 1e-6), 1e-3)
+
+        return pytest.approx(expected, **kwargs)
+
+    def allclose(x, y, **kwargs):
+        '''Same as: x == approx(y, **kwargs)'''
+        return x == approx(y, **kwargs)
+
+    def make_temp_file(*args, **kwargs):
+        '''Create a temporary file name'''
+        import os
+        from tempfile import mkstemp
+        fd, name = mkstemp(*args, **kwargs)
+        os.close(fd)
+        return name
+
+    ## Pytest options
+    @pytest.fixture(params=ti.supported_archs(), ids=ti.core.arch_name)
+    def taichi_archs(request):
+        marker = request.node.get_closest_marker('taichi')
+        req_arch = request.param
+
+        def ti_init(arch=[], exclude=[], require=[], **options):
```

Comment:
nit: might want to avoid immutable lists as default parameters, this works fine but just triggers some worry to me that it might cause some potential issues in some complicated cases. (e.g. we could just use Nones and a few conditions to substitute this)